src/traces/heatmap/clean_2d_array.js

```diff
@@ -30,12 +30,23 @@ module.exports = function clean2dArray(zOld, transpose) {
         old2new = function(zOld, i, j) { return zOld[i][j]; };
     }
 
+    var xMap = function(i) {return i;};
+    var yMap = function(i) {return i;};
+    if(trace && trace.type !== 'carpet' && trace.type !== 'contourcarpet') {
+        if(ya && ya.type === 'category') {
+            if(trace._y.length) yMap = function(i) {return trace._y[i];};
+        }
```

Comment:
Minor style tip: the above three line could become:

```js
      if(ya && ya.type === 'category' && trace.y._length) {
          yMap = function(i) { return trace._y[i]; };
      }
```

Reply:
Done in 59edb5704f076b438e6b8460b3e03097d61748ab